### PR TITLE
Move node configuration to runtime

### PR DESCRIPTION
**Pull request by tiziano88**
The ApplicationConfiguration object is now more of a directory of
executable wasm modules, and an initial Node configuration.

Subsequent Nodes are created at runtime by passing the serialized
NodeConfiguration protobuf message to `node_create` directly.

This allows removing an extra indirection from application to the config
of individual nodes, in favour of inlining configs in the Wasm code of
the main node itself.

Also remove parts of the logic around the old ApplicationConfiguration
format from the C++ Oak Runtime, under the assumption that it is going
to be deleted soon.

Ref #688

# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have written tests that cover the code changes.
  - [x] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [x] I have updated [documentation](docs/) accordingly.

### CODE REVIEW

**Comment by tiziano88** [comment]
Note that I have not fixed the glue code, and therefore some tests are currently failing.

My intention is to submit this only after we switch to using the Rust `oak_loader`, and in particular when the gRPC server pseudo-node stops being created in a special way by the Runtime before the actual entrypoint.

**Comment by tiziano88** [comment]
(but please do review in its current state, and let me know if you have any suggestions on whether I should go ahead and fix `oak_glue` right away -- maybe it is simpler than I think?)

File: examples/aggregator/module/rust/src/lib.rs [hunk 1]
```diff
@@ -92,7 +95,14 @@ impl AggregatorNode {
             bucket, svec
         );
 
-        match oak::grpc::client::Client::new("grpc-client", "").map(AggregatorClient) {
+        match oak::grpc::client::Client::new(&NodeConfiguration {
+            name: "grpc_client".to_string(),
+            config_type: Some(ConfigType::GrpcClientConfig(GrpcClientConfiguration {
+                address: "127.0.0.1:8888".to_string(),
+            })),
+        })
```
Comment (daviddrysdale) [inline examples/aggregator/module/rust/src/lib.rs:103]: Having helpers in the SDK for configuring standard pseudo-Node types would be helpful across all of the examples, e.g. so this could look something like:
```Rust
        match oak::grpc::client::Client::new(&oak::node_config::grpc_client("127.0.0.1:888"))
        .map(AggregatorClient)
        {
```

File: examples/hello_world/config/BUILD [hunk 2]
```diff
@@ -45,5 +45,5 @@ serialized_config(
         "app": "//:target/wasm32-unknown-unknown/release/hello_world.wasm",
         "translator": "//:target/wasm32-unknown-unknown/release/translator.wasm",
     },
-    textproto = ":config_translator.textproto",
+    textproto = ":config.textproto",
```
Comment (daviddrysdale) [inline examples/hello_world/config/BUILD:48]: Why the change here?

File: examples/hello_world/config/config_translator.textproto [hunk 3]
```diff
@@ -1,19 +0,0 @@
-node_configs {
```
Comment (daviddrysdale) [inline examples/hello_world/config/config_translator.textproto:1]: Why is this removed?  It seems to remove the ability to run `hello_world` with a `translator` module configured.

File: oak/server/rust/oak_runtime/src/config.rs [hunk 4]
```diff
@@ -14,121 +14,38 @@
...
-                    .and_then(|address| check_port(&address).map(|_| address))
-                    .map(|address| node::Configuration::GrpcServerNode { address })
-                    .map_err(|error| {
-                        error!("Incorrect gRPC server address: {:?}", error);
-                        OakStatus::ErrInvalidArgs
-                    })?,
-                Some(ConfigType::WasmConfig(WebAssemblyConfiguration { module_bytes, .. })) => {
-                    load_wasm(&module_bytes).map_err(|e| {
-                        error!("Error loading Wasm module: {}", e);
-                        OakStatus::ErrInvalidArgs
```
Comment (daviddrysdale) [inline oak/server/rust/oak_runtime/src/config.rs:106]: I like not having the data structures that mirror the protobuf config structure (and I think this would fix #755 along the way); are the validity checks (`check_port`, `load_wasm`) now done at a later point?

[Edit: yes they are]

File: oak/server/rust/oak_runtime/src/node/mod.rs [hunk 5]
```diff
@@ -86,63 +63,53 @@ impl std::fmt::Display for ConfigurationError {
...
+            runtime,
+            initial_reader,
+        )?)),
+        Some(ConfigType::GrpcClientConfig(_config)) => Ok(Box::new(external::PseudoNode::new(
+            name,
+            runtime,
+            initial_reader,
+        ))),
+        Some(ConfigType::RoughtimeClientConfig(_config)) => Ok(Box::new(
+            external::PseudoNode::new(name, runtime, initial_reader),
```
Comment (daviddrysdale) [inline oak/server/rust/oak_runtime/src/node/mod.rs:106]: Yeah, this will need to pass a copy of the node config over the Rust->C++ FFI boundary.  As you say, probably easiest to block this until `main()` flips into Rust and `oak_glue` gets reworked.

File: oak/server/rust/oak_abi/src/lib.rs [hunk 6]
```diff
@@ -119,31 +119,25 @@ extern "C" {
...
     ///
     /// Returns the status of the operation, as an [`OakStatus`] value.
     ///
     /// [`OakStatus`]: crate::OakStatus
+    /// [`NodeConfiguration`]: crate::proto::oak::application::NodeConfiguration
     pub fn node_create(
         config_buf: *const u8,
         config_len: usize,
-        entrypoint_buf: *const u8,
-        entrypoint_len: usize,
```
Comment (daviddrysdale) [inline oak/server/rust/oak_abi/src/lib.rs:146]: Also need to update `sdk/rust/oak/src/stubs.rs`

**Comment by tiziano88** [comment]
Thanks @daviddrysdale , yes I agree that this is indeed removing one layer of protection, but I think this is a layer that is not Oak's goal, and even if it is there, cannot really be leveraged in a meaningful way in its current form.

At least, I don't see a path forward in which this extra forced declaration (having the application declare upfront the kind of nodes it uses locally) could be made useful to clients. There is no mechanism for clients to inspect the configuration of a running application; even if there were, the application configuration is in a way an imprecise upper bound on the kind of capabilities that the application may rely on, and on the other hand it is not even good at that, because it does not enforce transitivity (e.g. if the application has a single egress gRPC client node, it may be tempting to conclude that it does not have storage, but the other side of the gRPC connection may actually be an application that has storage, etc.) this is the reason why we chose DIFC, which allows to reason about constraints in a distributed system. 

Even if we wanted to limit the functionality of an application to have some reassurance against local untrusted code, I think the current mechanism couples configuration and constraints too much; if we wanted to disable gRPC connectivity entirely, I think an explicit top-level bool in the application configuration may be a more intentional way of expressing that. The problem again is that it would still not really propagate across applications, or to storage, or anything else.

This is similar to the discussion on read / write rights on handles vs relying on information flow: the point of Oak is not to be a pure capability-based OS, it is to enforce policies in distributed systems, and sometimes being a pure capability system is at odd with that, or at the very least it does not add value (except perhaps some very localized value that does not extend to the rest of the system). But it's also certainly not impossible that some theory can unify both approaches at some point.

In the end, the way I see these things resolved is that effectively, client labels *are* capabilities of the kind of nodes that may manipulate the data; for instance, in order to allow storage, data would need to be labelled with a hypothetical storage pseudo-node secrecy label, which would allow such data to flow there (and not to other nodes); it is as if the data comes along with the capability to unlock that particular node type. It is a bit more nuanced than that, but perhaps this captures your intuition of some of this being in the spirit of capability-oriented systems.

WDYT?

File: examples/aggregator/module/rust/src/lib.rs (hunk 1 above)
Comment (tiziano88) [reply examples/aggregator/module/rust/src/lib.rs]: Nice idea, done.

File: examples/hello_world/config/BUILD (hunk 2 above)
Comment (tiziano88) [reply examples/hello_world/config/BUILD:48]: I deleted the other config, since now they are identical. I can keep both if you prefer though, let me know.

File: oak/proto/application.proto [hunk 7]
```diff
@@ -22,27 +22,34 @@ package oak.application;
...
-  // unique across the collection).  Each Node in the application will run under
-  // a configuration that is identified by an entry in this collection.
-  repeated NodeConfiguration node_configs = 1;
-  // Indication of what configuration the initial Node should run.  Must identify a
-  // NodeConfiguration entry that holds a WebAssemblyConfiguration object.
-  string initial_node_config_name = 2;
-  // The name of an exported Web Assembly function in the initial Node to
-  // be used as the Node's main entrypoint.
-  string initial_entrypoint_name = 3;
+  // Map from Wasm module names to their bytecode representation.
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/application.proto:29]: Done.

File: oak/proto/application.proto [hunk 8]
```diff
@@ -22,27 +22,34 @@ package oak.application;
...
+  //
+  // See https://webassembly.org/docs/binary-encoding/ .
+  //
+  // Entries in this map may be references by instances of WebAssemblyConfiguration
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/application.proto]: Done.

File: oak/proto/application.proto [hunk 9]
```diff
@@ -22,27 +22,34 @@ package oak.application;
...
+  // objects.
+  map<string, bytes> wasm_modules = 1;
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/application.proto:37]: Yes maps are just syntactic sugar for a repeated message generated by the proto compiler, they have to dedicated wire representation.

https://developers.google.com/protocol-buffers/docs/proto3#backwards-compatibility

File: oak/proto/application.proto [hunk 10]
```diff
@@ -22,27 +22,34 @@ package oak.application;
...
+
+  // Indication of what configuration the initial Node should run.
+  //
+  // Usually a NodeConfiguration entry that holds a WebAssemblyConfiguration object.
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/application.proto:41]: It may be a log node or a gRPC client node for some weird application I guess? Technically the runtime would just pipe through whatever you want.

File: oak/server/rust/oak_abi/src/lib.rs (hunk 6 above)
Comment (tiziano88) [reply oak/server/rust/oak_abi/src/lib.rs:146]: Done.

File: oak/server/rust/oak_runtime/src/config.rs (hunk 4 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/config.rs:106]: Affirmative.

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs [hunk 11]
```diff
@@ -140,65 +147,41 @@ impl WasmInterface {
         (interface, handle)
     }
 
-    /// Corresponds to the host ABI function [`node_create: (usize, usize, usize, usize,
-    /// u64) -> u32`](oak_abi::node_create).
+    /// Corresponds to the host ABI function [`node_create: (usize, usize, usize, usize, u64) ->
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/node/wasm/mod.rs]: Done.

File: oak/common/app_config_serializer.cc [hunk 12]
```diff
@@ -41,51 +41,49 @@ int main(int argc, char* argv[]) {
...
-        OAK_LOG(FATAL) << "Module path for " << module_name << " is not specified";
-      }
+    std::string module_name = module_info[0];
+    std::string module_path = module_info[1];
+    std::string module_bytes = oak::utils::read_file(module_path);
+    if (module_bytes.empty()) {
+      OAK_LOG(FATAL) << "Empty Wasm module:" << module_name;
+      return 1;
     }
+    (*config->mutable_wasm_modules())[module_name] = std::move(module_bytes);
```
Comment (ipetr0v) [inline oak/common/app_config_serializer.cc:80]: Can `[module_name]` return `nullptr`?

File: oak/server/rust/oak_runtime/src/node/grpc_server.rs [hunk 13]
```diff
@@ -94,18 +95,20 @@ impl GrpcServerNode {
...
-        config_name: &str,
+        display_name: &str,
+        config: GrpcServerConfiguration,
         runtime: RuntimeProxy,
-        address: SocketAddr,
         initial_reader: Handle,
-    ) -> Self {
-        Self {
-            config_name: config_name.to_string(),
+    ) -> Result<Self, ConfigurationError> {
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc_server.rs:102]: Is `new` supposed to return `Result` instead of just `Self`.
Should this function be called `create` instead? WDYT?

File: oak/server/rust/oak_runtime/src/node/grpc_server.rs [hunk 14]
```diff
@@ -23,23 +33,18 @@ use std::{
...
-};
-
-use crate::{metrics::METRICS, pretty_name_for_thread, runtime::RuntimeProxy, Handle};
-
 /// Struct that represents a gRPC server pseudo-Node.
 ///
 /// For each gRPC request from a client, gRPC server pseudo-Node creates a pair of temporary
 /// channels (to write a request to and to read a response from) and passes corresponding handles to
 /// the [`GrpcServerNode::channel_writer`].
+#[derive(Clone)]
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc_server.rs:41]: Why do we need `Clone` ?

File: oak/server/rust/oak_runtime/src/node/logger.rs [hunk 15]
```diff
@@ -14,34 +14,33 @@
...
 use prost::Message;
+use std::{
+    fmt::{self, Display, Formatter},
+    string::String,
+    thread::{self, JoinHandle},
+};
 
 pub struct LogNode {
-    config_name: String,
+    name: String,
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/logger.rs:34]: Should this field be called `display_name` (just like in `GrpcServerNode` and `WasmNode`) ?

File: examples/aggregator/module/rust/src/lib.rs [hunk 16]
```diff
@@ -92,7 +92,9 @@ impl AggregatorNode {
             bucket, svec
         );
 
-        match oak::grpc::client::Client::new("grpc-client", "").map(AggregatorClient) {
+        match oak::grpc::client::Client::new(&oak::node_config::grpc_client("127.0.0.1:8888"))
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/aggregator/module/rust/src/lib.rs:95]: I guess we will have root CAs pinned in the `oak_loader` binary, yes. I am not sure how this works currently though.

File: oak/common/app_config.cc [hunk 17]
```diff
@@ -34,23 +34,21 @@ constexpr int16_t kDefaultGrpcPort = 8080;
...
-constexpr char kGrpcClientConfigName[] = "grpc-client";
+constexpr char kAppWasmModuleName[] = "oak_module";
+constexpr char kAppWasmEntrypointName[] = "oak_main";
 
 }  // namespace
 
 std::unique_ptr<ApplicationConfiguration> DefaultConfig(const std::string& module_bytes) {
   auto config = absl::make_unique<ApplicationConfiguration>();
   config->set_grpc_port(kDefaultGrpcPort);
+  (*config->mutable_wasm_modules())[kAppWasmModuleName] = module_bytes;
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/common/app_config.cc:45]: No, my understanding is that this will create the entry if it's not there (it seems to work so I guess that's fine?)

Let me know if I'm doing it wrong though

File: oak/common/app_config_serializer.cc (hunk 12 above)
Comment (tiziano88) [reply oak/common/app_config_serializer.cc:80]: Hopefully it should be obvious if it does, anyways the size of `module_info` was checked above already. Or are you suggesting something else?

File: oak/server/rust/oak_runtime/src/node/grpc_server.rs (hunk 14 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/node/grpc_server.rs:41]: That's a question for you I guess :) I just removed the manual implementation and replaced it with `derive`.

File: oak/server/rust/oak_runtime/src/node/grpc_server.rs (hunk 13 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/node/grpc_server.rs:102]: I am also not quite sure about naming, but I think given the return type is obvious, it should be fine either way.

File: oak/server/rust/oak_runtime/src/node/logger.rs (hunk 15 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/node/logger.rs:34]: Done.

File: examples/abitest/config/config.textproto [hunk 18]
```diff
@@ -1,53 +1,8 @@
...
-  }
-}
-node_configs {
-  name: "absent-storage"
-  storage_config {
-    address: "test.invalid:9999"
-  }
-}
-node_configs {
-  name: "grpc-client"
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/abitest/config/config.textproto:30]: They are now in the code of `abitest` module itself!

**Comment by tiziano88** [comment]
@ipetr0v @daviddrysdale this is now ready for review.

Note that I removed parts of the logic around the old ApplicationConfiguration
format from the C++ Oak Runtime, under the assumption that it is going
to be deleted soon.

File: examples/abitest/tests/Cargo.toml [hunk 19]
```diff
@@ -15,6 +15,9 @@ log = "*"
 oak = "=0.1.0"
 oak_abi = "=0.1.0"
 prost = "*"
+# Using an old version that is supported by `cargo-raze`:
+# https://github.com/google/cargo-raze/issues/41#issuecomment-592274128
+tonic = { version = "=0.1.1", features = ["tls"] }
```
Comment (ipetr0v) [inline examples/abitest/tests/Cargo.toml:20]: I think after https://github.com/project-oak/oak/pull/1016 we can use the latest version of `tonic`

Comment (tiziano88) [reply examples/abitest/tests/Cargo.toml]: Done.

File: docs/programming-oak.md [hunk 20]
```diff
@@ -219,38 +219,21 @@ be serialized into a binary file. The Application first needs to specify a
...
-}
-grpc_port: 8080
-initial_node_config_name: "app"
-initial_entrypoint_name: "grpc_oak_main"
 ```
 <!-- prettier-ignore-end -->
 
-The `module_bytes: "<bytes>"` means that this value will be filled with
-WebAssembly module bytes after serialization using the
+The `wasm_modules` field (not shown above because empty in the template
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply docs/programming-oak.md]: Done. (I thought my version was grammatically correct though, was it not?)

File: docs/programming-oak.md [hunk 21]
```diff
@@ -515,20 +499,32 @@ that makes use of the functionality of the Oak Runtime.
...
+        oak_runtime::GrpcConfiguration {
+            grpc_server_tls_identity: None,
+            // Some of the tests require a gRPC client, so we populate the required certificate with
+            // an invalid value here, even though it will still fail when instantiating the actual
+            // gRPC client.
+            grpc_client_root_tls_certificate: Some(Certificate::from_pem("invalid-cert")),
+        },
+    )
+    .expect("unable to configure runtime with test wasm");
 ```
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply docs/programming-oak.md:566]: Both done.

File: oak/proto/application.proto [hunk 22]
```diff
@@ -22,27 +22,33 @@ package oak.application;
 //
 // An Oak Application is built from a collection of interconnected Nodes,
 // each of which is running the code described by an entry in this
-// configuration.  These Nodes are created dynamically at runtime, with
+// configuration. These Nodes are created dynamically at runtime, with
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/application.proto]: Reworded.

File: oak/server/rust/oak_runtime/src/node/mod.rs [hunk 23]
```diff
@@ -108,112 +69,65 @@ impl std::fmt::Display for ConfigurationError {
                 write!(f, "Failed to parse an address: {}", e)
             }
             ConfigurationError::IncorrectPort => write!(f, "Incorrect port (must be > 1023)"),
+            ConfigurationError::IncorrectURI => write!(f, "Incorrect URI"),
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/node/mod.rs:73]: It is only logged at the point where the parsing happens. I start thinking that we should probably just either log or propagate all errors, there is not much value in this halfway split. Created https://github.com/project-oak/oak/issues/1027 to track.

File: sdk/rust/oak_tests/src/lib.rs [hunk 24]
```diff
@@ -76,23 +80,33 @@ pub fn run_single_module(
...
-        configuration,
+        application_configuration,
         oak_runtime::RuntimeConfiguration::default(),
-        oak_runtime::GrpcConfiguration::default(),
+        oak_runtime::GrpcConfiguration {
+            grpc_server_tls_identity: None,
+            // Some of the tests require a gRPC client, so we populate the required certificate with
+            // an invalid value here, even though it will still fail when instantiating the actual
+            // gRPC client.
+            grpc_client_root_tls_certificate: Some(Certificate::from_pem("invalid-cert")),
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply sdk/rust/oak_tests/src/lib.rs:102]: Because it would cause it to fail too early, for some of the tests (they would not be able to create the gRPC client at all otherwise).

File: oak/server/rust/oak_runtime/src/node/mod.rs [hunk 25]
```diff
@@ -32,8 +29,8 @@ mod storage;
 mod wasm;
 
 /// Trait encapsulating execution of a Node or pseudo-Node.
-pub trait Node {
-    /// Execute the Node, using the provided `Runtime` reference and initial handle. The method
+pub trait Node: Send {
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/node/mod.rs:32]: It was already needed, but it was specified as an additional type bound on the actual `node_create` function (which was quite confusing), I just moved it here instead.

File: sdk/rust/oak_tests/Cargo.toml [hunk 26]
```diff
@@ -7,11 +7,14 @@ license = "Apache-2.0"
...
 lazy_static = "*"
 log = { version = "*", features = ["std"] }
 oak = "=0.1.0"
 oak_abi = "=0.1.0"
 oak_runtime = { version = "=0.1.0", features = ["test_build"] }
 prost = "*"
 rand = { version = "*" }
-cargo_metadata = "*"
+# Using an old version that is supported by `cargo-raze`:
+# https://github.com/google/cargo-raze/issues/41#issuecomment-592274128
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply sdk/rust/oak_tests/Cargo.toml]: Done.
